Use gson instead of the jackson library for json operations

pfeodrippe opened the pull request:
Fixes #29.

It replaces the jackson dependencies by the gson one so we have
a more compatible license (see issue link above).

It's a very direct conversion, I've even maintained method names like
`getArrayNode` or `getObjectNode`, it does not appear to be worth to
change it to follow the implementation classes names.

pfeodrippe commented:
I'll check later if I need to add the gson license.

lemmy commented:
TLC includes other licenses such as [`jline-LICENSE.txt`](https://github.com/tlaplus/tlaplus/tree/master/tlatools/org.lamport.tlatools/lib/jline) in its jar file.

lemmy commented on README.md:
```diff
@@ -12,23 +12,25 @@ The Modules
 How to use it
 -------------
 
+You must be running Java 9 or higher.
```
Why Java 9?  Gson seems to work fine with Java 1.8 here.

pfeodrippe replied:
The problem is at `IOUtils.java` which uses `readAllBytes`, it was introduced only at Java 9, but maybe I'm configuring the JVM in a invalid way? Let me know.

Running
```shell
$ export JAVA_HOME=/usr/lib/jvm/java-1.8.0-openjdk-amd64
$ ant test
```

returns

```java
...
compile:
    [javac] Compiling 9 source files to /home/rafael/dev/CommunityModules/build/modules
    [javac] /home/rafael/dev/CommunityModules/modules/tlc2/overrides/IOUtils.java:89: error: cannot find symbol
    [javac] 		final StringValue stdout = new StringValue(new String(process.getInputStream().readAllBytes()));
    [javac] 		                                                                              ^
    [javac]   symbol:   method readAllBytes()
    [javac]   location: class InputStream
    [javac] /home/rafael/dev/CommunityModules/modules/tlc2/overrides/IOUtils.java:90: error: cannot find symbol
    [javac] 		final StringValue stderr = new StringValue(new String(process.getErrorStream().readAllBytes()));
    [javac] 		                                                                              ^
    [javac]   symbol:   method readAllBytes()
    [javac]   location: class InputStream
    [javac] /home/rafael/dev/CommunityModules/modules/tlc2/overrides/IOUtils.java:130: error: cannot find symbol
    [javac] 		final StringValue stdout = new StringValue(new String(process.getInputStream().readAllBytes()));
    [javac] 		                                                                              ^
    [javac]   symbol:   method readAllBytes()
    [javac]   location: class InputStream
    [javac] /home/rafael/dev/CommunityModules/modules/tlc2/overrides/IOUtils.java:131: error: cannot find symbol
    [javac] 		final StringValue stderr = new StringValue(new String(process.getErrorStream().readAllBytes()));
    [javac] 		                                                                              ^
    [javac]   symbol:   method readAllBytes()
    [javac]   location: class InputStream
    [javac] 4 errors
```

lemmy replied:
Oh, good catch!  Can you please open a new issue for IOUtils and another one to configure the GitHub action to run with 1.8?

pfeodrippe replied:
Opened both issues, thanks!

lemmy commented on modules/tlc2/overrides/Json.java:
```diff
@@ -345,114 +339,119 @@ private static JsonNode getArrayNode(IValue value) throws IOException {
     return new RecordValue(keys.toArray(new UniqueString[0]), values.toArray(new Value[0]), false);
   }
 
-
```
Please mark the `resolve` method and its counterpart in `TLCOverride` for removal after graduation to TLC.

pfeodrippe replied:
Added a @Deprecated annotation, thanks!

lemmy commented on README.md:
```diff
@@ -37,6 +39,14 @@ Alternatively, you can download a [library archive](https://github.com/tlaplus/C
 Being a community-driven repository puts the community in charge to check the validity and correctness of submissions. The maintainers of this repository will try to keep this place in order, but we can't guarantee the quality of the
 modules and therefore cannot provide any assistance on eventual malfunctions.
 
+Test
```
This should probably be moved below the Contributing section.

pfeodrippe replied:
Yes, I've thought that you may test before contributing (or that they don't have a specific order), buy I've added it below the Contributing section, thanks!

lemmy commented on modules/Json.tla:
```diff
@@ -86,6 +86,13 @@ JsonDeserialize(absoluteFilename) ==
   (*************************************************************************)
   CHOOSE val : TRUE
 
+JsonDeserializeWithNull(absoluteFilename, nullValue) ==
```
What is `nullValue` for?

pfeodrippe replied:
I've learned that in TLA+ we don't have the concept of null, right? Tell me if I'm mistaken (or if something new was introduced which allows it).

When you use a deserialization operator, you may want to handle nulls. When it's not handled, it just returns garbage (like `%2%` in my terminal). So I've thought that the user may have a `null` representation which he whishes to use, like a default value (or replacement for nulls). See example below.

```ruby
# Let's say we have a `my-json.json` file with the contents `[{"amount": 10}, {"amount": null}]`
# We will run some assertions below (all are `TRUE`)
<<[amount |-> 10], [amount |-> 0]>>      = JsonDeserializeWithNull("my-json.json", 0)

<<[amount |-> 10], [amount |-> "null"]>> = JsonDeserializeWithNull("my-json.json", "null")

<<[amount |-> 10], [amount |-> <<>>]>>   = JsonDeserializeWithNull("my-json.json", <<>>)
```

There is a test which uses the fixture file (have a better name for this type of file?) `json-with-null.json` where I replace the default value with a set.

pfeodrippe replied:
I will add comments to the tla and java files if these operators are desired, but I can delete too if you if you feel like it =D

lemmy replied:
I now realize that you only propose it for *de*-serialize (indicating that the ergonomics of the `Json*` operator names is not ideal).  Do you have a concrete use-case in mind where the Json input has `null` values with which we could validate the `*WithNull` operator? For example, we could also consider passing in a `LAMBDA` to give users more control.  At any rate, I suggest moving the `*WithNull` enhancement to a separate issue.

pfeodrippe replied:
Got it, I am gonna do that later in the day. Thanks. 

The point is that we return null value from the deserialization, but it does not work correctly. Maybe we could not handle Json null values for now until we have a solution.

pfeodrippe replied:
Removed.

pfeodrippe commented:
> TLC includes other licenses such as jline-LICENSE.txt in its jar file. Anyway, the plan is to "graduate" the Json module into TLC. Until then, it should be good enough to replace the Jackson jars with Gson's jar in CommunityModules-deps.jar.

Thanks o/ I have already replaced everything from Jackson with Gson (including the jars) and now added the Gson license.

lemmy commented:
Merged with minor adjustments in https://github.com/tlaplus/CommunityModules/commit/90f3562874e8a8916bba025705573bb37a9e7283. Thanks!
